### PR TITLE
Physics engine improvements

### PR DESCRIPTION
**Pull request by hamzzam**
## Summary

Physics engine improvements to get physics in a state more suitable for state sync feature.
This includes enabling physics on the server-side. Enable local physics for all objects whether owned or not, just the final state of objects is based on the incoming network state. This allows for a more natural flow of things since there is no lag in a player experiencing the physics effects of their actions.

Also included a system to generate physics simulations for testing purposes. This system is in a somewhat rough state.

## Checklist
- [ ] Pre-push checks pass `npm run check`
  - [ ] Linter passing via `npm run lint`
  - [ ] Unit & Integration tests passing via `npm run test:packages`
  - [ ] Docker build process passing via `npm run build-client`
- [ ] If this PR is still a WIP, convert to a draft 
- [ ] When this PR is ready, mark it as "Ready for review"
- [ ] Changes have been manually QA'd
- [ ] Changes reviewed by at least 2 approved reviewers

## References

_References to pertaining issue(s)_

## QA Steps

1. `git checkout pr_branch_name`
2. `npm install`
3. `npm run dev-reinit`
4. `npm run dev`

List any additional steps required to QA the changes of this PR, as well as any supplemental images or videos.

## Reviewers

@HexaField @speigg

### CODE REVIEW

**Comment by hamzzam** [comment]
Thanks @zulqarnainhanif 😄

File: packages/engine/src/physics/systems/PhysicsSystem.ts [hunk 1]
```diff
@@ -116,20 +97,35 @@ export default async function PhysicsSystem(world: World) {
...
+      const transform = getComponent(entity, TransformComponent)
+      const body = collider.body as PhysX.PxRigidDynamic
+
+      teleportRigidbody(body, transform.position, transform.rotation)
+    }
+
     for (const entity of colliderQuery()) {
+      // Hackish fix for now to disable some physics effects in editor
+      // We should prolly not register the physics system when loading editor
+      if (Engine.isEditor) return
```
Comment (HexaField) [inline packages/engine/src/physics/systems/PhysicsSystem.ts:116]: `return` will take us out of the entire ECS loop here, if we want to exit the for loop, use `continue`. as per the style guide, we also want to take all these loops out of the main system function into their own module scope functions as a more composable style of programming

File: packages/engine/src/physics/systems/PhysicsSystem.ts [hunk 2]
```diff
@@ -79,24 +78,6 @@ export default async function PhysicsSystem(world: World) {
   await world.physics.createScene()
 
   return () => {
```
Comment (HexaField) [inline packages/engine/src/physics/systems/PhysicsSystem.ts:80]: what might be wise here, is to move all of the enter and exit queries to the start of the system loop, and move all the simulation logic to the end of the system loop, and before the simulation logic with an `if (Engine.isEditor) return`

File: packages/projects/default-project/PhysicsSimulationTestSystem.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  // TODO: Generating the object as a network object here may result in an error sometimes,
+  // the reason being the spawn object network action from server is received before the object has been added to the scene locally.
+  // In order to test network functionality, call the generateSimulationData directly from scene loading logic in code for now.
+  generatePhysicsObject(config, new Vector3(0, 15, 0))
+}
+
+const defaultSpawnPosition = new Vector3()
+export const generatePhysicsObject = (
```
Comment (HexaField) [inline packages/projects/default-project/PhysicsSimulationTestSystem.ts:100]: this is all great, we should pull bits of this out and use it in unit tests for the physics system

**Review by HexaField — CHANGES_REQUESTED** [review]

File: packages/projects/default-project/PhysicsSimulationTestSystem.ts (hunk 3 above)
Comment (hamzzam) [reply packages/projects/default-project/PhysicsSimulationTestSystem.ts:100]: Will do in the tests.

File: packages/engine/src/physics/systems/PhysicsSystem.ts [hunk 4]
```diff
@@ -55,144 +63,137 @@ function updateBoundingBox(entity: Entity, force = false) {
...
+    for (const entity of colliderQuery.exit()) {
+      const colliderComponent = getComponent(entity, ColliderComponent, true)
+      if (colliderComponent?.body) {
+        world.physics.removeBody(colliderComponent.body)
       }
     }
 
-    // clear collision queue
-    world.physics.collisionEventQueue = []
+    processRaycasts(world.physics)
```
Comment (HexaField) [inline packages/engine/src/physics/systems/PhysicsSystem.ts:180]: let's pass `world` into all of these, and use a pipe

**Review by HexaField — CHANGES_REQUESTED** [review]

**Review by HexaField — APPROVED** [review]

**Pull request merged**